### PR TITLE
Added "fail faster" support. Closes #587

### LINKED ISSUE #587: Support 'Fail Fast'
In cases where call caching is not being used (as is the case for Genomes On the Cloud for the time being), the ability to have workflows "fail fast" is desired.  Fail fast in this case mean, once a task has failed no new tasks should be started (currently running tasks can finish) and then the workflow should be set to failed.

This is important because when running a workflow, and you can't reuse the non-failed work, you want to know the workflow will fail as soon as possible.  With a long running workflow, an early failure with few downstream dependencies you have to wait a long time (24 hours for GOTC) to realize it's actually failed.

This should be specifyable as a workflow option, e.g. failure-strategy = fast where the current behavior is 'slow'.  If it's easy to also have this as a server config setting for the default that would be a nice to have but not critical.

### PR DESCRIPTION
**Pull request by cjllanwarne**
Suggested comments for reviewers:
- [x] Needs a test
- [x] Needs docs
- [x] Add to changelog
- [x] How about a proper type for this?
- [x] Best code I've ever seen
- [x] The test is failing

### CODE REVIEW

File: engine/src/main/scala/cromwell/engine/WorkflowDescriptor.scala [hunk 1]
```diff
@@ -40,6 +40,7 @@ case class WorkflowDescriptor(id: WorkflowId,
                               backend: Backend,
                               configCallCaching: Boolean,
                               lookupDockerHash: Boolean,
+                              configWorkflowFailureMode: Boolean,
```
Comment (mcovarr) [inline engine/src/main/scala/cromwell/engine/WorkflowDescriptor.scala]: How about a proper type for this?  Not just because types are teh awesome but also as your title suggests it's not a stretch to imagine a "fail fastest" that aborts whatever's still running.

Comment (cjllanwarne) [inline engine/src/main/scala/cromwell/engine/WorkflowDescriptor.scala]: This wasn't on the approved comments list!

File: engine/src/main/scala/cromwell/engine/workflow/WorkflowActor.scala [hunk 2]
```diff
@@ -825,10 +826,16 @@ case class WorkflowActor(workflow: WorkflowDescriptor)
       case _ => false
     }
 
+    lazy val hasAnythingFailed = executionStore exists {
```
Comment (mcovarr) [inline engine/src/main/scala/cromwell/engine/workflow/WorkflowActor.scala]: `executionStore.values exists { x => x == Failed || x == Aborted }`

**Comment by mcovarr** [comment]
Consider all those boxes checked.

File: engine/src/main/scala/cromwell/engine/WorkflowDescriptor.scala (hunk 1 above)
Comment (geoffjentry) [inline engine/src/main/scala/cromwell/engine/WorkflowDescriptor.scala]: @mcovarr knows how to make me :smile:

Comment (cjllanwarne) [inline engine/src/main/scala/cromwell/engine/WorkflowDescriptor.scala]: OTOH, agile is all about incremental changes and not over-engineering solutions for a "maybe, possibly..." future scenario :)

**Comment by mcovarr** [comment]
The one Travis failure in "push" occurs in the "fail slow" test :frowning:

**Comment by cjllanwarne** [comment]
@mcovarr all suggested comments addressed. Ready for round 2 now

**Comment by cjllanwarne** [comment]
The wheel also chooses @Horneth

File: CHANGELOG.md [hunk 3]
```diff
@@ -2,6 +2,9 @@
 
 ## 0.19
 
+* Workflow options now allows you to specify a `workflowFailureMode` to control workflow behaviour after a call has failed, for example:`{ "workflowFailureMode": "..." }`. The options are:
+  * `ContinueWhilePossible` - continues to start and process calls in the workflow, as long as they did not depend on the failing call
+  * `NoNewCalls` - no *new* calls are started but existing calls are allowed to finish
```
Comment (mcovarr) [inline CHANGELOG.md]: nice!

File: CHANGELOG.md [hunk 4]
```diff
@@ -2,6 +2,9 @@
 
 ## 0.19
 
+* Workflow options now allows you to specify a `workflowFailureMode` to control workflow behaviour after a call has failed, for example:`{ "workflowFailureMode": "..." }`. The options are:
```
Comment (mcovarr) [inline CHANGELOG.md]: behaviour -> behavior

File: CHANGELOG.md (hunk 3 above)
Comment (mcovarr) [inline CHANGELOG.md]: What's the default behavior?

File: engine/src/test/scala/cromwell/util/SampleWdl.scala [hunk 5]
```diff
@@ -2317,7 +2317,7 @@ task shouldCompleteFast {
         |    Int a
         |    command {
         |        echo "The number was: ${a}"
-        |        sleep 0.2
+        |        sleep 0.15
```
Comment (mcovarr) [inline engine/src/test/scala/cromwell/util/SampleWdl.scala]: why these changes?

File: engine/src/main/scala/cromwell/engine/WorkflowDescriptor.scala [hunk 6]
```diff
@@ -61,6 +62,7 @@ case class WorkflowDescriptor(id: WorkflowId,
     log.warn(s"$CallLogsDirOptionKey expected to be of type String", e)
     throw e
   }
+  val workflowFailureMode: WorkflowFailureMode = workflowOptions.get(WorkflowFailureModeKey) flatMap {WorkflowFailureMode.tryParse} getOrElse configWorkflowFailureMode
```
Comment (mcovarr) [inline engine/src/main/scala/cromwell/engine/WorkflowDescriptor.scala]: It looks as if `WorkflowFailureMode` is specified incorrectly this would silently ignore that and just go with what's in the config?

File: engine/src/test/scala/cromwell/WorkflowFailSlowSpec.scala [hunk 7]
```diff
@@ -5,13 +5,39 @@ import cromwell.engine.WorkflowFailed
...
+      |}
+    """.stripMargin
+
+  val FailFastOptions =
+    """
+      |{
+      |  "workflowFailureMode": "NoNewCalls"
+      |}
+    """.stripMargin
+
```
Comment (mcovarr) [inline engine/src/test/scala/cromwell/WorkflowFailSlowSpec.scala]: add a test for a bad value

Comment (mcovarr) [inline engine/src/test/scala/cromwell/WorkflowFailSlowSpec.scala]: and a test for no value

File: README.md [hunk 8]
```diff
@@ -975,18 +975,23 @@ Example workflow options file:
...
+    * **outputs_path** - Specifies a path where final workflow outputs will be written.  If this is not specified, workflow outputs will not be copied out of the Cromwell workflow execution directory/path.
+    * **call_logs_dir** - Specifies a path where final call logs will be written.  If this is not specified, call logs will not be copied out of the Cromwell workflow execution directory/path.
+    * **defaultRuntimeOptions** - A JSON object where the keys are [runtime attributes](#runtime-attributes) and the values are defaults that will be used through the workflow invocation.  Individual tasks can choose to override these values.  See the [runtime attributes](#specifying-default-values) section for more information.
+    * **workflowFailureMode** - What happens after a task fails. Choose from:
+        * **ContinueWhilePossible** - continues to start and process calls in the workflow, as long as they did not depend on the failing call
+        * **NoNewCalls** - no *new* calls are started but existing calls are allowed to finish
```
Comment (mcovarr) [inline README.md]: doc default behavior here too

Comment (mcovarr) [inline README.md]: shouldn't there be a bit about the config setting in here somewhere?

File: engine/src/main/scala/cromwell/engine/workflow/WorkflowActor.scala [hunk 9]
```diff
@@ -825,10 +826,14 @@ case class WorkflowActor(workflow: WorkflowDescriptor)
       case _ => false
     }
 
+    lazy val hasAnythingFailed = executionStore.values exists { x => x == Failed || x == Aborted }
+
+    lazy val workflowFailFastTrigger = workflowFailureMode == NoNewCalls && hasAnythingFailed
```
Comment (mcovarr) [inline engine/src/main/scala/cromwell/engine/workflow/WorkflowActor.scala]: it feels like most of this logic should live in the `WorkflowFailureMode` subtypes

File: engine/src/test/scala/cromwell/util/SampleWdl.scala (hunk 5 above)
Comment (cjllanwarne) [inline engine/src/test/scala/cromwell/util/SampleWdl.scala]: Turns out Fail Fast-ish lets there be a race condition here, if the failing and successful tasks take the same time, sometimes the successful task's successor has already started by the time the failure is noticed. So this increases (from 0) the gap between the preceding failure and success.

Comment (mcovarr) [inline engine/src/test/scala/cromwell/util/SampleWdl.scala]: Can you please leave a comment to this effect?

File: README.md [hunk 10]
```diff
@@ -975,18 +975,23 @@ Example workflow options file:
...
+* JES Backend Only
+    * **jes_gcs_root** - (JES backend only) Specifies where outputs of the workflow will be written.  Expects this to be a GCS URL (e.g. `gs://my-bucket/workflows`).  If this is not set, this defaults to the value within `backend.jes.baseExecutionBucket` in the [configuration](#configuring-cromwell).
```
Comment (Horneth) [inline README.md]: Totally off PR but I'm thinking about it as I see it now. The "root" concept is actually also totally valid for local backend and this option should probably be generalized.

File: engine/src/main/scala/cromwell/engine/WorkflowDescriptor.scala [hunk 11]
```diff
@@ -299,8 +302,8 @@ object WorkflowDescriptor {
     val validatedDescriptor = for {
       c <- validateCoercedInputs(id, rawInputs, namespace).disjunction
       d <- validateDeclarations(id, namespace, options, c, engineFunctions).disjunction
-    } yield WorkflowDescriptor(id, sourceFiles, options, workflowLogOptions(conf), rawInputs, namespace, c, d, backend, configCallCaching(conf), lookupDockerHash(conf),
-      wfContext, fileSystems)
+    } yield new WorkflowDescriptor(id, sourceFiles, options, workflowLogOptions(conf), rawInputs, namespace, c, d, backend, configCallCaching(conf), lookupDockerHash(conf),
```
Comment (Horneth) [inline engine/src/main/scala/cromwell/engine/WorkflowDescriptor.scala]: just wondering, why the `new` ?

File: engine/src/test/scala/cromwell/WorkflowFailSlowSpec.scala (hunk 7 above)
Comment (cjllanwarne) [inline engine/src/test/scala/cromwell/WorkflowFailSlowSpec.scala]: No test added. Bad value test added to centaur

File: README.md (hunk 8 above)
Comment (cjllanwarne) [inline README.md]: Hmm, apparently the config is "self-documenting" so I've mentioned it here and the config has some more explanation.

File: engine/src/main/scala/cromwell/engine/workflow/WorkflowActor.scala (hunk 9 above)
Comment (cjllanwarne) [inline engine/src/main/scala/cromwell/engine/workflow/WorkflowActor.scala]: Hmm, I guess I can move the behaviour definition inside the object. The implementation of that behaviour itself though, I think belongs here

File: engine/src/main/scala/cromwell/engine/WorkflowDescriptor.scala [hunk 12]
```diff
@@ -40,6 +40,7 @@ case class WorkflowDescriptor(id: WorkflowId,
                               backend: Backend,
                               configCallCaching: Boolean,
                               lookupDockerHash: Boolean,
+                              configWorkflowFailureMode: WorkflowFailureMode,
```
Comment (Horneth) [inline engine/src/main/scala/cromwell/engine/WorkflowDescriptor.scala]: I think that if this is going to be an attribute of the class it should be the final definitive value.
Now it is used as a default when computing the actual value on line 65.
The full logic of determining the mode should be done int eh apply IMO. It would avoid having `configWorkflowFailureMode` and `workflowFailureMode` where really you only care about the latter.

File: engine/src/main/scala/cromwell/engine/workflow/WorkflowActor.scala (hunk 9 above)
Comment (Horneth) [inline engine/src/main/scala/cromwell/engine/workflow/WorkflowActor.scala]: Hmm I don't see why ? I can't see how we would have different classes that implement this behaviour differently in Cromwell ?
there could be a 
`def shouldFail(hasAnythingFailed: Boolean)` method in `WorkflowFailureMode` that would return `hasAnythingFailed` for `NoNewCalls` and `false` for `ContinueWhilePossible`

Comment (mcovarr) [inline engine/src/main/scala/cromwell/engine/workflow/WorkflowActor.scala]: @Horneth yes that's what I meant, thanks.  "most of this logic" was not a good way for me to describe that.  :smile:

File: engine/src/main/scala/cromwell/engine/WorkflowDescriptor.scala (hunk 12 above)
Comment (cjllanwarne) [inline engine/src/main/scala/cromwell/engine/WorkflowDescriptor.scala]: This is actually in a slightly weird spot w.r.t validation, defaults, etc. I'm going to wait until @gauravs90 commits the ValidationActor stuff because it's all going to be so much neater in the brave new world.

**Comment by Horneth** [comment]
:+1:

**Comment by mcovarr** [comment]
looks like a test is failing for realz

File: engine/src/main/scala/cromwell/engine/WorkflowDescriptor.scala [hunk 13]
```diff
@@ -61,6 +62,8 @@ case class WorkflowDescriptor(id: WorkflowId,
     log.warn(s"$CallLogsDirOptionKey expected to be of type String", e)
     throw e
   }
+  // This exception should be validated in ValidationActor
```
Comment (mcovarr) [inline engine/src/main/scala/cromwell/engine/WorkflowDescriptor.scala]: it's called something else now

File: engine/src/main/scala/cromwell/engine/workflow/WorkflowActor.scala [hunk 14]
```diff
@@ -825,10 +826,14 @@ case class WorkflowActor(workflow: WorkflowDescriptor)
       case _ => false
     }
 
+    lazy val hasAnythingFailed = executionStore.values exists { x => x == Failed || x == Aborted }
+
+    lazy val disallowNewCalls = !workflowFailureMode.allowNewCallsAfterFailure && hasAnythingFailed
```
Comment (mcovarr) [inline engine/src/main/scala/cromwell/engine/workflow/WorkflowActor.scala]: maybe flip the operands

File: engine/src/main/scala/cromwell/engine/workflow/WorkflowActor.scala [hunk 15]
```diff
@@ -825,10 +826,14 @@ case class WorkflowActor(workflow: WorkflowDescriptor)
...
+
     def isRunnable(entry: ExecutionStoreEntry) = {
       entry match {
-        case (key: FinalCallKey, ExecutionStatus.NotStarted) => nonFinalCallsComplete
-        case (key, ExecutionStatus.NotStarted) => arePrerequisitesDone(key)
+        case (key: FinalCallKey, ExecutionStatus.NotStarted) => nonFinalCallsComplete && !disallowNewCalls
+        case (key, ExecutionStatus.NotStarted) => arePrerequisitesDone(key) && !disallowNewCalls
```
Comment (mcovarr) [inline engine/src/main/scala/cromwell/engine/workflow/WorkflowActor.scala]: can this be rephrased positively rather than doubly negative?

File: engine/src/main/scala/cromwell/engine/workflow/WorkflowActor.scala (hunk 14 above)
Comment (mcovarr) [inline engine/src/main/scala/cromwell/engine/workflow/WorkflowActor.scala]: eh scratch that, it's probably more efficient the way you've got it

File: engine/src/main/scala/cromwell/engine/workflow/WorkflowActor.scala (hunk 15 above)
Comment (cjllanwarne) [inline engine/src/main/scala/cromwell/engine/workflow/WorkflowActor.scala]: `I don't see why not`

**Comment by Horneth** [comment]
Do we want this for the release ?

File: engine/src/main/scala/cromwell/engine/backend/WorkflowDescriptor.scala [hunk 16]
```diff
@@ -61,6 +62,8 @@ case class WorkflowDescriptor(id: WorkflowId,
     log.warn(s"$CallLogsDirOptionKey expected to be of type String", e)
     throw e
   }
+  // This exception should be validated in ValidationActor
```
Comment (mcovarr) [inline engine/src/main/scala/cromwell/engine/backend/WorkflowDescriptor.scala]: The actor formerly known as ValidationActor

Comment (geoffjentry) [inline engine/src/main/scala/cromwell/engine/backend/WorkflowDescriptor.scala]: nay, née

Comment (cjllanwarne) [inline engine/src/main/scala/cromwell/engine/backend/WorkflowDescriptor.scala]: I haven't rebased this yet! Give me a chance!

File: engine/src/test/scala/cromwell/util/SampleWdl.scala [hunk 17]
```diff
@@ -2306,7 +2306,8 @@ task shouldCompleteFast {
         |    Int a
         |    command {
         |        echo "The number was: ${a}"
-        |        sleep 0.2
+        |        # More than 0.15 so this should finish second
```
Comment (mcovarr) [inline engine/src/test/scala/cromwell/util/SampleWdl.scala]: This 0.1 second stuff still makes me kind of uneasy, wouldn't be surprised if Travis gets wonky with this

**Comment by mcovarr** [comment]
:+1: assuming Travis approves

**Comment by mcovarr** [comment]
unhappy Travis :cry:
